### PR TITLE
Milestone:Ruby Solo Project

### PR DESCRIPTION
**Pull request by shubhsk88**
This PR contains the implementations of the define methods in an array using the module functionality

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
   DisplayCopNames: true
 
 Layout/LineLength:
   Max: 120
 Metrics/MethodLength:
-  Max: 20
+  Max: 30
 Metrics/AbcSize:
-  Max: 50
+  Max: 100
```
Comment (shshamim63) [inline .rubocop.yml]: You should not alter the default configuration file. If you believe that default configuration is not enough please open an issue in this [link](https://github.com/microverseinc/linters-config/tree/master/ruby). If you think that some rules you want to disable for now you can use the inline disabled.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# Array-Method Ruby
```
Comment (shshamim63) [inline README.md]: They are enumerable methods. Not array methods. So you can name it more specific to the task that you have covered in here.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Array-Method Ruby
+
+This Project showcases and implements the Array method of ruby using the Enumerable,The method represented in this project as follow
```
Comment (shshamim63) [inline README.md]: Did you mean something like this ->
`The project contains ruby enumerable methods. Where all the methods interact with an array and return the same value as the original method does. It covers the following enumerable methods.`

File: method.rb [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      end
+
+    else
+      my_each do |num|
+        new_arr.push(yield num)
+      end
+    end
+
+    new_arr
+  end
```
Comment (shshamim63) [inline method.rb]: [Optional]
I just tried to refactor the code in the following way,
```ruby
def my_map(prop = nil)
    return to_enum(:my_map) unless block_given?
    new_arr = []
    my_each do |num|
      new_arr.push(prop ? prop.call(num) : yield(num))
    end
    new_arr
end
```

File: method.rb [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      acc = yield(acc, num)
+    end
+    acc
+  end
+
+  def multiply_els
+    my_inject(1) do |acc, num|
+      acc * num
+    end
+  end
```
Comment (shshamim63) [inline method.rb]: multiply_els is not an enumerable method, so it should be placed outside of the module. And based on the instruction it should seem something like following ->
```ruby
def multiply_els array
  array.my_inject { |acc, num| acc * num }
end
```

File: method.rb [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    end
+    counter
+  end
+
+  def my_inject(acc = 0)
+    my_each do |num|
+      acc = yield(acc, num)
+    end
+    acc
+  end
```
Comment (shshamim63) [inline method.rb]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```

File: method.rb [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+  def my_count(counter = size)
+    p_count = 0
+    if block_given?
+      my_each do |num|
+        p_count += 1 if yield num
+      end
+      counter = p_count
+    end
+    counter
+  end
```
Comment (shshamim63) [inline method.rb]: `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- counts the number of items in enum that are equal to item if an argument is given
```ruby
[1, 1, 2 ,3].my_count(1)  #should return 2
```

File: method.rb [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      my_each do |num|
+        stat = false if yield num
+      end
+    else
+      my_each do |num|
+        stat = !num
+      end
+    end
+    stat
+  end
```
Comment (shshamim63) [inline method.rb]: `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2, 3].my_none?(String)  #should return true
['1', '2', '3'].my_none?(String)  #should return false
```
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/)   #should return true
%w[dog door rod blade].my_none?(/d/)   #should return false
```
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[1, 2, 4].my_none?(3) #should return true
[1, 2, 3].my_none?(3) #should return false
```

File: method.rb [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+        stat = true if yield num
+      end
+    else
+      my_each do |num|
+        bool = !num
+        stat = !bool
+      end
+    end
+    stat
+  end
```
Comment (shshamim63) [inline method.rb]: `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/t/)    #should return false.
%w[dog door rod blade].my_any?(/d/)   #should return true.
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
[1, 2, 3].my_any?(3)   #should return true
[1, 2, 2].my_any?(3)   #should return false
```

File: method.rb [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+    elsif block_given?
+      my_each do |num|
+        stat = false unless yield num
+      end
+    else
+      return stat
+    end
+    stat
+  end
```
Comment (shshamim63) [inline method.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
%w[dog door rod blade].my_all?(/t/)   #should return false.
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @shubhsk88 :wave: ,
amazing job :clap: well-indented code, returning enumerator based on the method name. Just found some issues, well they are easy to cover. Let's get to them and complete this project.
___
### :spades: Mandatory Changes :spades: 
___
1. The following methods are not covering some edge cases. Please check the comments attached to respective block of code.
- my_all?
- my_any?
- my_none?
- my_count
- my_inject
2. The multiply_els method is not placed on the proper position and needs to accept an array as a parameter. Check the comment.
3. You can brush up the readme, check out the comments.
4. You should not change the default configuration of linter files. Please check the comment also.
___
### :clubs: Optional Changes :clubs: 
___
I have refactored your my_map method. You can also refactor other methods too.
___
##### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

File: method.rb [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+          stat = num.to_s.match?(arg.to_s)
+        end
+      end
+
+    elsif block_given?
+      my_each do |num|
+        stat = false unless yield num
+      end
+    else
+      my_each do |num|
```
Comment (uryelah) [inline method.rb:60]: - [ ] When you don't give a block or argument to the my_all? method it should return true only if all of the enumerable items are 'truthy'.
So an array with a 'falsy' item should return false wit this method. [Read this article if you are not sure which are the falsy and truthy values in ruby.](https://andycroll.com/ruby/truthiness-in-conditionals/)
Fix the code inside your your else conditional to fix this.

File: method.rb [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    new_arr
+  end
+
+  def my_all?(arg = nil) # rubocop:disable Style/MethodLength
+    stat = true
+
+    if arg
+      if arg.class == Class
+        my_each do |num|
+          stat = num.is_a?(arg)
```
Comment (uryelah) [inline method.rb:44]: - [ ] The stat should be true only if all of the items in the enumerable are of the given Class.
Make sure it returns false the first time if finds something that is not from the given Class.

File: method.rb [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+      end
+    end
+    stat
+  end
+
+  def my_any?(arg = nil) # rubocop:disable Style/MethodLength
+    stat = false
+
+    if arg
+      if arg.class == Class
```
Comment (uryelah) [inline method.rb:71]: - [ ] Make sure the my_any also works when a pattern other than a Class or Regexp is given.
EX:
[0,1,'ok', [], false, 'lol'].any?('lol')
=> true

File: method.rb [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    end
+    stat
+  end
+
+  def my_none?(arg = nil) # rubocop:disable Style/MethodLength
+    stat = true
+
+    if arg
+      if arg.class == Class
+        my_each do |num|
```
Comment (uryelah) [inline method.rb:103]: - [ ] Here you have the opposite issue you have on the my_all? method.
It should return false as soon as you find any item that is from the argument Class.
Ex:
[1,2,3,'work', true].none?(TrueClass)
=> false

File: method.rb [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+          stat = !num.is_a?(arg)
+        end
+      else
+        my_each do |num|
+          stat = !num.to_s.match?(arg.to_s)
+        end
```
Comment (uryelah) [inline method.rb:111]: - [ ] Remember to add a condition for a pattern other than a Class or a Regex.

File: method.rb [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+      end
+
+    end
+    my_each do |num|
+      p_count += 1 if num == counter
+    end
+    p_count
+  end
+
+  def my_inject(acc = 0, sign = nil)
```
Comment (uryelah) [inline method.rb]: - [ ] my_inject should also be able to take a block.
Make sure it works with for example an array or strings and a block that adds only the strings longer than 3 characters.

File: method.rb [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+  end
+end
+
+def multiply_els(array)
+  array.my_inject { |acc, num| acc * num }
+end
+
+# arr = [4, 6, 3]
+
+# p arr.my_all?{|num| num%2==0}
```
Comment (uryelah) [inline method.rb]: - [ ] Make sure to remove all of your commented out code before requesting a ode review.

**Review by uryelah — CHANGES_REQUESTED** [review]
Hello **Shubham Singh**, I'm Sarah, your reviewer for this PR 👩‍🚀  

**First of all, good job with your code. It improved a lot from the first review 🌟**

There are still some issues to fix though. [Make sure you read the specifications of the enumerable module](https://ruby-doc.org/core-2.7.1/Enumerable.html) again with care and check the required changes in the comments below. You can do it! After you are done with this your ruby foundation will be solid as a rock!💪 

---------     

## Required Changes 🏗 👷‍♀ 

----

**Comment by uryelah** [comment]
**Kindly fix the REQUIRED issues and send another code review request.**

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**:+1: for all your efforts on this project. You have just a small case left to fix**
- **my_any?** - when no block or argument is given, my_any? should return true **only** if at least one of the elements is true (that is not false or nil).
e.g `[false, nil].my_any?` should return `false` and not `true`, since none of the elements is true.


_**Kindly fix this and submit another code review request.**_

**Cheers:stars:**

### Status: Changes Requested.

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :clap: :clap: :clap: 
Hi there @shubhsk88 :hand: 
Your worked a lot on this project and should be proud of it :tada: :+1: 
You can merge the PR now. :smiley: 
Good luck with the next one

**Happy coding :computer:** 

**Sergio Torres**

**Pull request merged**